### PR TITLE
fix issue with MVP case indicators blowing away form indicators

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?165274

@dannyroberts has the most context for this. @TylerSheffels @esoergel and @nickpell have also all worked on this ticket. fix is in https://github.com/dimagi/commcare-hq/commit/3bcffc2c9f98dcc60eb23e2216f497bc0ceba970 rest is tests and docs.

wait for build

### CODE REVIEW

File: custom/_legacy/mvp_docs/tests/test_pillows.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
 import json
 import os
 from couchdbkit import ResourceNotFound, ResourceConflict
-import datetime
-from django.test.testcases import TestCase
+from django.test.testcases import TestCase, SimpleTestCase
```
Comment (dimagimon) [inline custom/_legacy/mvp_docs/tests/test_pillows.py]: F401 'SimpleTestCase' imported but unused

File: custom/_legacy/mvp_docs/tests/test_pillows.py [hunk 2]
```diff
@@ -123,3 +119,51 @@ def test_delete_doc_that_doesnt_exist(self):
...
+        self.assertEqual(29, len(updated_computed))
+        self.assertEqual('child_visit_form', computed['child_visit_form']['value'])
+
+
+def _fake_indicators(filename):
+    with open(os.path.join(os.path.dirname(__file__), 'data', filename)) as f:
+        indicators = json.loads(f.read())
+        return [_wrap(i) for i in indicators]
+
+def _wrap(indicator):
```
Comment (dimagimon) [inline custom/_legacy/mvp_docs/tests/test_pillows.py]: E302 expected 2 blank lines, found 1

**Comment by czue** [comment]
this needs both MVP pillows to be reset (I think). Possibly just the form one.

File: custom/_legacy/mvp_docs/tests/data/mvp-sauri-case-form-indicators.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[
```
Comment (czue) [inline custom/_legacy/mvp_docs/tests/data/mvp-sauri-case-form-indicators.json]: pulled all these from the prod database

File: custom/_legacy/mvp_docs/pillows.py [hunk 4]
```diff
@@ -149,17 +149,22 @@ def process_indicators(self, namespaces, domain, doc_dict):
...
             except ResourceNotFound:
-                pillow_eval_logging.error(
-                    "Could not find an XFormInstance with id %(xform_id)s "
-                    "related to Case %(case_id)s" % {
-                        'xform_id': xform_id,
-                        'case_id': doc_dict['_id'],
-                    }
-                )
-                continue
+                # if that fails fall back to the main DB
```
Comment (dannyroberts) [inline custom/_legacy/mvp_docs/pillows.py]: Doesn't this mean if there's a race condition, this will revert to the old, failing, behavior?

Comment (dannyroberts) [inline custom/_legacy/mvp_docs/pillows.py]: And by "if there's a race condition" I mean that in the race condition between the form and the case pillows, if the case pillow processes a case before the form pillow processes all of its forms.

Comment (czue) [inline custom/_legacy/mvp_docs/pillows.py]: hmm, yeah it does.

I took this route (as opposed to loading up all the indicators) for performance reasons (many indicators are slow and load other documents), but you're right there are problems with it.

let me think about it / tweak it a bit. sorry was just excited that it worked.

Comment (dannyroberts) [inline custom/_legacy/mvp_docs/pillows.py]: Haha no worries. I can understand the excitement!

**Comment by dannyroberts** [comment]
re: resetting MVP pillows—heads up that last time it took a long time (maybe a week?) for it to catch up, and a lot of back and forth between MVP and Nate ("Nope, still not caught up! It'll be another day at least"). At some point it'll probably be worth writing an mvp fast reindex, since I'm pretty sure a big part of the delay is simply the fact the changes feed is so slow...

**Comment by dannyroberts** [comment]
(I could be totally off-base about that)

File: custom/_legacy/mvp_docs/tests/test_pillows.py [hunk 5]
```diff
@@ -123,3 +118,81 @@ def test_delete_doc_that_doesnt_exist(self):
...
+            MVPFormIndicatorPillow().change_transform(form_json)
+            updated_form = IndicatorXForm.get(form_json['_id'])
+            updated_computed = updated_form.computed_['mvp_indicators']
+            self.assertEqual(29, len(updated_computed))
+            self.assertEqual('child_visit_form', updated_computed['child_visit_form']['value'])
+
+        self._call_with_patches(_test)
+
+
+    @patch('corehq.apps.indicators.utils.get_namespaces')
```
Comment (dimagimon) [inline custom/_legacy/mvp_docs/tests/test_pillows.py]: E303 too many blank lines (2)

File: custom/_legacy/mvp_docs/tests/test_pillows.py [hunk 6]
```diff
@@ -123,3 +118,81 @@ def test_delete_doc_that_doesnt_exist(self):
...
+    @patch('corehq.apps.indicators.models.FormIndicatorDefinition.get_all')
+    def _call_with_patches(self, fn, form_get_all_patch, case_get_all_patch,
+                           case_form_get_all_patch, get_namespaces_patch):
+        form_get_all_patch.return_value = _fake_indicators('mvp-sauri-form-indicators.json')
+        case_get_all_patch.return_value = _fake_indicators('mvp-sauri-case-indicators.json')
+        case_form_get_all_patch.return_value = _fake_indicators('mvp-sauri-case-form-indicators.json')
+        get_namespaces_patch.return_value = ['mvp_indicators']
+        fn()
+
+def _save_form_and_case():
```
Comment (dimagimon) [inline custom/_legacy/mvp_docs/tests/test_pillows.py]: E302 expected 2 blank lines, found 1

File: custom/_legacy/mvp_docs/pillows.py (hunk 4 above)
Comment (czue) [inline custom/_legacy/mvp_docs/pillows.py]: updated so that if it doesn't find the doc it computes all the indicators. does that seem right?

**Comment by czue** [comment]
agree on fast reindex. also am going to add some caching to the indicators

**Comment by czue** [comment]
(can do in a separate PR though)

**Comment by czue** [comment]
looks like travis is behind but tests passed after the main change and I ran the `mvp_docs` tests after. would be nice if this gets in today. cc @TylerSheffels

**Comment by czue** [comment]
@orangejenny if this gets merged we need to reset the two MVP pillows. I think @esoergel maybe just did something similar and can help you with it

**Comment by TylerSheffels** [comment]
Good by me, @dannyroberts because you understand the change better

**Comment by esoergel** [comment]
I think Danny's offline for the rest of the day.  @orangejenny hit me up if you want me to walk you through resetting the pillows.

File: custom/_legacy/mvp/README.md [hunk 7]
```diff
@@ -30,6 +33,37 @@ If your project has the slug `mvp-sauri` (this is what comes after
...
+1. The `change_transform` function handles deletions and then calls `process_indicators`
+2. All `FormIndicatorDefinition` objects for the form's `(domain, xmlns)` combination are pulled out
+3. Those `FormIndicatorDefinition`s are applied to the form in bulk
+4. The form is saved back to the new database.
+
+## Case processing
+
+Case processing is very similar to form processing except there is a final step. All form-dependent indicators are applied to all the case's forms each time the case is processed. This means that the case processing can update indicators in forms!
+
+
```
Comment (benrudolph) [inline custom/_legacy/mvp/README.md]: :+1:

**Comment by dannyroberts** [comment]
@czue @biyeun re: fast reindex; I just came across a mgmt cmd called `mvp_force_update`. Haven't gone deeply into it, but wanted to flag in case that is basically what we want

**Comment by czue** [comment]
thanks for the pointer - i'll check it out.

**Pull request merged**